refactor(experimental): update RPC documentation in library README (#2336)
solana-labs/solana-web3.js: packages/library/README.md

web3.js 2.0 ships with an implementation of the [JSON RPC specification](https://www.jsonrpc.org/specification) and a type spec for the [Solana JSON RPC](https://docs.solana.com/api).

The main package responsible for managing communication with an RPC is `@solana/rpc`. However, this package makes use of more granular packages to break down the RPC logic into smaller pieces. Namely, these packages are:

-   `@solana/rpc`: Contains all logic related to sending Solana RPC calls.
-   `@solana/rpc-api`: Describes all Solana RPC methods using types.
-   `@solana/rpc-transport-http`: Provides a concrete implementation of an RPC transport using HTTP requests.

> -   `@solana/rpc-spec`: Defines the JSON RPC spec for sending RPC calls using proxies.

Reviewer: The proxy thing is sort of incidental. Nobody needs to know that, and also _part_ of the sender is concrete code (ie. the param/response transforms) so it's not _completely_ a proxy.

-   `@solana/rpc-spec-types`: Shared JSON RPC specifications types and helpers that are used by both `@solana/rpc` and `@solana/rpc-subscriptions` (described in the next section).
-   `@solana/rpc-types`: Shared Solana RPC types and helpers that are used by both `@solana/rpc` and `@solana/rpc-subscriptions`.

That being said, the main `@solana/web3.js` library re-exports the `@solana/rpc` package so, going forward, we will import RPC types and functions from the library directly.

### Getting Started with RPC Calls

To get started with RPC calls, you may use the `createSolanaRpc` function by providing the URL of the Solana JSON RPC server. This will create a default client for interacting with the Solana JSON RPC.

```tsx
import { createSolanaRpc } from '@solana/web3.js';

// Create an RPC client.
const rpc = createSolanaRpc('http://127.0.0.1:8899');

> //    ^ Rpc<SolanaRpcApi>

Reviewer: Maybe use the actual TypeScript playground syntax, for those in the know.
```suggestion
//    ^? Rpc<SolanaRpcApi>
```

    address: string;
    circulatingSupply: number;
    description: string;
    erc721: boolean;
    erc1155: boolean;
    genesisBlock: string;
    genesisTransaction: string;
    name: string;
    totalSupply: number;
}>;

// Set up an interface for the request method.
interface NftCollectionDetailsApi {
    // Define the method's name, parameters and response type
    qn_fetchNFTCollectionDetails(args: { contracts: string[] }): NftCollectionDetailsApiResponse;
}

// Export the type spec for downstream users.
export type QuickNodeRpcApi = NftCollectionDetailsApi;
```

Here’s how a developer might use it:

```tsx
import { createDefaultRpcTransport, createRpc, createRpcApi } from '@solana/web3.js';

// Create the custom API.
const api = createRpcApi<QuickNodeRpcApi>();

// Set up an HTTP transport.
const transport = createDefaultRpcTransport({ url: 'http://127.0.0.1:8899' });

// Create the RPC client.
const quickNodeRpc = createRpc({ api, transport });
//    ^ Rpc<QuickNodeRpcApi>
```

As long as a particular JSON RPC method adheres to the [official JSON RPC specification](https://www.jsonrpc.org/specification), it will be supported by web3.js 2.0.

> ### Aborting RPC Requests

Reviewer: Thanks for bringing this back; I was worried for a second :)

RPC requests are now abortable with modern `AbortControllers`. When calling an RPC method such as `getSlot`, it will return a `PendingRpcRequest` proxy object that contains a `send` method to send the request to the server.

```ts
const pendingRequest: PendingRpcRequest<Slot> = rpc.getSlot();

const slot: Slot = await pendingRequest.send();
```

Since the arguments of the `getSlot` method are reserved for the request payload, the `send` method is the one that can accept additional arguments such as an `AbortSignal` to abort the request.

> Here's an example of a custom `AbortController` used to abort a request after a 5-second timeout:

Reviewer: I'm _slightly_ worried that people will think this is only useful for timeouts, and not – for example – in cases where your UI unmounts, or an operation rolls back. I guess IYKYK.

Author: Good shout, I changed the example to something that doesn't use timeout and added another line of documentation:

--- 

[...]

Aborting RPC requests can be useful for a variety of things such as setting a timeout on a request or cancelling a request when a user navigates away from a page.

```tsx
import { createSolanaRpc } from '@solana/web3.js';

const rpc = createSolanaRpc('http://127.0.0.1:8900');

// Create a new AbortController.
const abortController = new AbortController();

// Abort the request when the user navigates away from the current page.
function onUserNavigateAway() {
    abortController.abort();
}

// The request will be aborted if and only if the user navigates away from the page.
const slot = await rpc.getSlot().send({ abortSignal: abortController.signal });
```

```tsx
import { createSolanaRpc } from '@solana/web3.js';

const rpc = createSolanaRpc('http://127.0.0.1:8900');

> // Create a new AbortController such that it aborts after 5 seconds.
> const abortController = new AbortController();
> setTimeout(() => {
>     abortController.abort();
> }, 5000);
> 
> // Get the latest slot, but abort the request after 5 seconds if it hasn't completed.
> const slot = await rpc.getSlot().send({ abortSignal: abortController.signal });

Reviewer: ```suggestion
// Get the latest slot, but abort the request after 5 seconds if it hasn't completed.
const slot = await rpc.getSlot().send({ abortSignal: AbortSignal.timeout(5000) });
```

Author: See https://github.com/solana-labs/solana-web3.js/pull/2336#discussion_r1530056855

} catch (e) {
    // The subscription went down.
    // Retry it and then recover from potentially having missed
    // a balance update, here (eg. by making a `getBalance()` call).
}
```

You can read more about `AsyncIterator` at the following links:

-   [Mozilla Developer Docs: `AsyncIterator`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/AsyncIterator)
-   [Luciano Mammino (Blog): JavaScript Async Iterators](https://www.nodejsdesignpatterns.com/blog/javascript-async-iterators/)

### Aborting RPC Subscriptions

Similarly to RPC calls, applications can terminate active subscriptions using an `AbortController` attribute on the `subscribe` method. In fact, this parameter is _required_ for subscriptions to encourage you to clean up subscriptions that your application no longer needs.

Let's take a look at some concrete examples that demonstrate how to abort subscriptions.

#### Subscription Timeout

Here's an example of an `AbortController` used to abort a subscription after a 5-second timeout:

```tsx
import { createSolanaRpcSubscriptions } from '@solana/web3.js';

const rpcSubscriptions = createSolanaRpcSubscriptions('ws://127.0.0.1:8900');

> // Create a new AbortController.
> const abortController = new AbortController();
> 
> // Subscribe for slot notifications using the AbortController's signal.
> const slotNotifications = await rpcSubscriptions.slotNotifications().subscribe({ abortSignal: abortController.signal });
> 
> // Set a timer for 5 seconds, then abort the controller.
> setTimeout(() => {
>     abortController.abort();
> }, 5000);

Reviewer: ```suggestion
// Subscribe for slot notifications using the AbortController's signal.
const slotNotifications = await rpcSubscriptions.slotNotifications().subscribe({ abortSignal: AbortSignal.timeout(5000) });
```

const rpcSubscriptions = createSolanaRpcSubscriptionsFromTransport(transport);
//    ^ RpcSubscriptions<SolanaRpcSubscriptionsApi>
```

### Scoping the RPC Subscriptions API

Using the `createSolanaRpcSubscriptions` or `createSolanaRpcSubscriptionsFromTransport` functions, we always get the same RPC Subscriptions API, including all Solana RPC stable subscriptions. However, since the RPC Subscriptions API is described using types only, it is possible to scope the API to a specific set of subscriptions and even add your own custom subscriptions.

#### Scoping by Cluster

If you're using a specific cluster, you may wrap your RPC URL inside a helper function like `mainnet` or `devnet` to inject that information into the RPC type system.

```ts
import { createSolanaRpcSubscriptions, mainnet, devnet } from '@solana/web3.js';

const mainnetRpc = createSolanaRpcSubscriptions(mainnet('https://api.mainnet-beta.solana.com'));
//    ^ RpcSubscriptionsMainnet<SolanaRpcSubscriptionsApi>

const devnetRpc = createSolanaRpcSubscriptions(devnet('https://api.devnet.solana.com'));
//    ^ RpcSubscriptionsDevnet<SolanaRpcSubscriptionsApi>
```

#### Including Unstable Subscriptions

> If your app needs access to [unstable RPC Subscriptions](https://docs.solana.com/api/websocket#blocksubscribe) — e.g. `BlockNotificationsApi` or `SlotsUpdatesNotificationsApi` — you may use the `createSolanaRpcSubscriptions_UNSTABLE` and `createSolanaRpcSubscriptionsFromTransport_UNSTABLE` functions to create an RPC Subscriptions client that includes those subscriptions.

Reviewer: Maybe mention ‘…and your RPC server supports them’
